core/checksums/src/main/java/software/amazon/awssdk/checksums/internal/ConstructorCache.java

```diff
@@ -63,6 +63,11 @@ private Optional<Class<?>> getClass(String className) {
                 return Optional.empty();
             }
         });
+        // if the WeakReference to the class has been garbage collected, remove it from the cache and try again
+        if (classRef.isPresent() && classRef.get().get() == null) {
+            classesByClassLoader.remove(classLoader);
+            return getClass(className);
+        }
         return classRef.map(WeakReference::get);
```

Comment:
Is it still possible for the weak ref to get GC'd between 67 and 71? Would it be better to store `classRef.get().get()` to a temporary var to avoid that happening?

Reply:
Ooff - yeah, good catch - I've updated.

On a related note - I had been focused on checking that this was thread safe - and because of the usage of synchronized maps, I believe it is - in particular the `classesByClassLoader.remove(classLoader)` and `classesByClassLoader.computeIfAbsent` generally ensure the cache is consistent.  There is an issue where a class could be removed and re-loaded multiple times which would result in extra work, but not in an inconsistent/incorrect cache state.